hack/generator/pkg/codegen/pipeline/unroll_recursive_types.go

```diff
@@ -0,0 +1,188 @@
+// }
+// This usually happens for Error types (CloudError and the like). The recursive references must be removed because
+// controller-tools doesn't support generating "references" (JSON $ref) so it can't support recursive types today.
+// See https://github.com/kubernetes-sigs/controller-tools/issues/489 for more information.
+// Unrolling these types, while required for controller-tools to function, means
+// THAT THE TYPES WE GENERATE DON'T EXACTLY CONFORM TO THE PUBLISHED API!
+// In practice, we think this is ok because the types that reference themselves only return a depth of 1.
+// If we were to unroll all loops (rather than just types that directly reference themselves like we're doing here) that
+// "in practice" observation may no longer hold, so we avoid doing it here (it's also more complicated to do that).
+// TODO: We could require you to name the type in config or code in order to get it unrolled? There are currently 43 types being unrolled by this
+func UnrollRecursiveTypes() Stage {
+	return MakeStage(
+		UnrollRecursiveTypesStageID,
+		"Unrolls directly recursive types, as they are not supported by controller-gen",
+		func(ctx context.Context, state *State) (*State, error) {
+
+			result := make(astmodel.Types)
+
+			// Find object types that reference themselves
+
+			fixer := newSimpleRecursiveTypeFixer()
+
+			for _, def := range state.Types() {
+				updatedDef, err := fixer.visitor.VisitDefinition(def, simpleRecursiveTypeFixerContext{name: def.Name()})
+				if err != nil {
+					return nil, err
+				}
+
+				result.Add(updatedDef)
+			}
+
+			// Add new unrolled types
+			result.AddTypes(fixer.newTypes)
+
+			return state.WithTypes(result), nil
+		})
+}
+
+// TODO: If we end up needing more complex cycle detection we can rework TypeWalker to do it, but for now this
+// TODO: is simpler and seems to cover all the cases we need.
+// simpleRecursiveTypeFixer removes circular references from types that refer to themselves directly.
+type simpleRecursiveTypeFixer struct {
+	visitor  *astmodel.TypeVisitor
+	newTypes astmodel.Types
+}
+
+type simpleRecursiveTypeFixerContext struct {
+	name         astmodel.TypeName
+	mustUnroll   bool
+	unrolledName astmodel.TypeName
+}
+
+func (c simpleRecursiveTypeFixerContext) WithUnrolledName(name astmodel.TypeName) simpleRecursiveTypeFixerContext {
+	c.unrolledName = name
+	c.mustUnroll = true
+	return c
+}
+
+func newSimpleRecursiveTypeFixer() *simpleRecursiveTypeFixer {
+	result := &simpleRecursiveTypeFixer{
+		newTypes: make(astmodel.Types),
+	}
+
+	visitor := astmodel.TypeVisitorBuilder{
+		VisitObjectType: astmodel.MakeIdentityVisitOfObjectTypeWithPerPropertyContext(result.UnrollObjectTypeProperty),
+		VisitTypeName:   result.UnrollRecursiveReference,
+	}.Build()
+	result.visitor = &visitor
+
+	return result
+}
+
+func (s *simpleRecursiveTypeFixer) UnrollObjectTypeProperty(ot *astmodel.ObjectType, prop *astmodel.PropertyDefinition, ctx interface{}) (interface{}, error) {
```

Comment:
Shouldn't this be private? It feels like part of the internal implementation. (Maybe you made it public so it could be called from `astmodel.MakeIdentityVisit...`? I'm _nearly_ sure you don't need to.)

Reply:
No you're right, I was being careless here. Good catch